// @threlte/extras/dist/hooks/useProgress.js
import { writable } from 'svelte/store';
import { DefaultLoadingManager } from 'three';
let previousTotalLoaded = 0;
export const useProgress = () => {
    const activeStore = writable(false);
    const itemStore = writable(undefined);
    const loadedStore = writable(0);
    const totalStore = writable(0);
    const errorsStore = writable([]);
    const progressStore = writable(0);
    DefaultLoadingManager.onStart = (item, loaded, total) => {
        activeStore.set(true);
        itemStore.set(item);
        loadedStore.set(loaded);
        totalStore.set(total);
        progressStore.set((loaded - previousTotalLoaded) / (total - previousTotalLoaded));
    };
    DefaultLoadingManager.onLoad = () => {
        activeStore.set(false);
    };
    DefaultLoadingManager.onError = (item) => {
        errorsStore.update((errors) => {
            return [...errors, item];
        });
    };
    DefaultLoadingManager.onProgress = (item, loaded, total) => {
        if (loaded === total) {
            previousTotalLoaded = total;
        }
        activeStore.set(true);
        itemStore.set(item);
        loadedStore.set(loaded);
        totalStore.set(total);
        progressStore.set((loaded - previousTotalLoaded) / (total - previousTotalLoaded) || 1);
    };
    return {
        active: { subscribe: activeStore.subscribe },
        item: { subscribe: itemStore.subscribe },
        loaded: { subscribe: loadedStore.subscribe },
        total: { subscribe: totalStore.subscribe },
        errors: { subscribe: errorsStore.subscribe },
        progress: { subscribe: progressStore.subscribe }
    };
};
